### PR TITLE
Add 10 additional cops for detected deprecated behavior

### LINKED ISSUE #79: Remove any attributes in the metadata.rb file
### What category of cop is this?:

Please try to select as few as possible.

- [x] Deprecation - A cop that alerts to a deprecation in the Chef Infra Client
- [ ] Correctness - A cop that alerts when incorrect coding behavior is being used
- [ ] Sharing - A cop that alerts to missing functionality necessary for sharing cookboooks with others
- [ ] Style - A cop that alerts to a style best practice
- [ ] Simplification - A cop that alerts to ways code can be simplified through refactoring
- [ ] Effortless - A cop that alerts on code that must be resolved to move to the Effortless pattern
- [ ] Policyfiles - A cop that alerts on code that must be resolved to move to the Policyfiles pattern

### Describe the new cop:
We no longer support documenting attributes in the metadata.rb file. We should remove these if we find them

### What it would trigger on:
```ruby
attribute 'accounts', display_name: 'Accounts Hash',
                      description: 'Hash of Accounts attributes',
                      type: 'hash'
```

### What it would autocorrect to if applicable
Just delete it

### LINKED ISSUE #80: Remove `provides` in the metadata.rb file
### What category of cop is this?:

Please try to select as few as possible.

- [x] Deprecation - A cop that alerts to a deprecation in the Chef Infra Client
- [ ] Correctness - A cop that alerts when incorrect coding behavior is being used
- [ ] Sharing - A cop that alerts to missing functionality necessary for sharing cookboooks with others
- [ ] Style - A cop that alerts to a style best practice
- [ ] Simplification - A cop that alerts to ways code can be simplified through refactoring
- [ ] Effortless - A cop that alerts on code that must be resolved to move to the Effortless pattern
- [ ] Policyfiles - A cop that alerts on code that must be resolved to move to the Policyfiles pattern

### Describe the new cop:
provides doesn't make sense in the metadata.rb file anymore. We should remove these.

### What it would trigger on:
```ruby
provides 'account::user'
```

### What it would autocorrect to if applicable
Just delete it

### LINKED ISSUE #85: Add a check for the easy_install resource
### What category of cop is this?:

Please try to select as few as possible.

- [x] Deprecation - A cop that alerts to a deprecation in the Chef Infra Client
- [ ] Correctness - A cop that alerts when incorrect coding behavior is being used
- [ ] Sharing - A cop that alerts to missing functionality necessary for sharing cookboooks with others
- [ ] Style - A cop that alerts to a style best practice
- [ ] Simplification - A cop that alerts to ways code can be simplified through refactoring
- [ ] Effortless - A cop that alerts on code that must be resolved to move to the Effortless pattern
- [ ] Policyfiles - A cop that alerts on code that must be resolved to move to the Policyfiles pattern

### Describe the new cop:
We removed this resource in Chef 13

### What it would trigger on:
Any usage of easy_install

### What it would autocorrect to if applicable
N/A

### LINKED ISSUE #91: Detect usage of erl_call resource
### What category of cop is this?:

Please try to select as few as possible.

- [x] Deprecation - A cop that alerts to a deprecation in the Chef Infra Client
- [ ] Correctness - A cop that alerts when incorrect coding behavior is being used
- [ ] Sharing - A cop that alerts to missing functionality necessary for sharing cookboooks with others
- [ ] Style - A cop that alerts to a style best practice
- [ ] Simplification - A cop that alerts to ways code can be simplified through refactoring
- [ ] Effortless - A cop that alerts on code that must be resolved to move to the Effortless pattern
- [ ] Policyfiles - A cop that alerts on code that must be resolved to move to the Policyfiles pattern

### Describe the new cop:
erl_call was removed in Chef 14

### What it would trigger on:
Any usage of the erl_call resource

### What it would autocorrect to if applicable
N/A

### LINKED ISSUE #106: Detect a chef_handler resource with the supports property
### What category of cop is this?:

Please try to select as few as possible.

- [x] Deprecation - A cop that alerts to a deprecation in the Chef Infra Client
- [ ] Correctness - A cop that alerts when incorrect coding behavior is being used
- [ ] Sharing - A cop that alerts to missing functionality necessary for sharing cookboooks with others
- [ ] Style - A cop that alerts to a style best practice
- [ ] Simplification - A cop that alerts to ways code can be simplified through refactoring
- [ ] Effortless - A cop that alerts on code that must be resolved to move to the Effortless pattern
- [ ] Policyfiles - A cop that alerts on code that must be resolved to move to the Policyfiles pattern

### Describe the new cop:
We renamed the supports property to type to avoid confusion. We should autocorrect this anywhere we find it.

### What it would trigger on:
```ruby
chef_handler 'whatever' do
  supports 'foo'
end
````

### What it would autocorrect to if applicable
```ruby
chef_handler 'whatever' do
  type 'foo'
end
````

See: https://github.com/chef/chef/blob/master/lib/chef/resource/chef_handler.rb#L47

### PR DESCRIPTION
**Pull request by tas50**
## Description
Add 10 additional cops for detected deprecated behavior

Chef/ConflictsMetadata:
  Description: Don't use the deprecated 'conflicts' metadata value

Chef/SuggestsMetadata:
  Description: Don't use the deprecated 'suggests' metadata value

Chef/ReplacesMetadata:
  Description: Don't use the deprecated 'replaces' metadata value

Chef/ProvidesMetadata:
  Description: Don't use the deprecated 'provides' metadata value

Chef/AttributeMetadata:
  Description: Don't use the deprecated 'attribute' metadata value

Chef/CookbookDependsOnCompatResource:
  Description: Don't depend on the deprecated compat_resource cookbook

Chef/CookbookDependsOnPartialSearch:
  Description: Don't depend on the deprecated partial_search cookbook

Chef/EasyInstallResource:
  Description: Don't use the deprecated easy_install resource

Chef/ErlCallResource:
  Description: Don't use the deprecated erl_call resource

UnnecessaryDependsChef14
  Description: Don't depend on cookbooks made obsolete by Chef 14

## Related Issue
Fixes #91 
Fixes #85
Fixes #80
Fixes #79

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Chore (non-breaking change that does not add functionality or fix an issue)

## Checklist:

- [x] I have read the **CONTRIBUTING** document.
- [ ] I have run the pre-merge tests locally and they pass.
- [x] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [x] All commits have been signed-off for [the Developer Certificate of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

### CODE REVIEW

File: config/cookstyle.yml [hunk 1]
```diff
@@ -135,6 +185,10 @@ Chef/WhyRunSupportedTrue:
   Description: why_run_supported? no longer needs to be set to true as it is the default in Chef 13+
   Enabled: true
   VersionAdded: '5.1.0'
+    # Include:
```
Comment (tyler-ball) [inline config/cookstyle.yml]: Is there a reason these are added commented out?

File: lib/rubocop/cop/chef/deprecation/easy_install.rb [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+      #
+      # @example
+      #
+      #   # bad
+      #   easy_install "my_thing" do
+      #     bar
+      #   end
+      #
+      class EasyInstallResource < Cop
+        MSG = "Don't use the deprecated easy_install resource".freeze
```
Comment (tyler-ball) [inline lib/rubocop/cop/chef/deprecation/easy_install.rb]: Do we have a suggested remedy for users who have this resource?

File: lib/rubocop/cop/chef/deprecation/suggests_metadata.rb [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+module RuboCop
+  module Cop
+    module Chef
+      # Make sure ignore_failure is used instead of epic_fail
+      #
+      # @example
+      #
+      #   # bad in metadata.rb:
+      #
+      #   replaces "another_cookbook"
```
Comment (tyler-ball) [inline lib/rubocop/cop/chef/deprecation/suggests_metadata.rb]: Should be `suggests`

File: lib/rubocop/cop/chef/modernize/chef_14_resources.rb [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+      #   depends 'mac_os_x'
+      #   depends 'chef_handler'
+      #   depends 'swap'
+      #   depends 'chef_hostname'
+      #
+      class UnnecessaryDependsChef14 < Cop
+        MSG = "Don't depend on cookbooks made obsolete by Chef 14".freeze
+
+        def_node_matcher :legacy_depends?, <<-PATTERN
+          (send nil? :depends (str {"build-essential" "sysctl" "mac_os_x" "chef_handler" "swap" "chef_hostname"}))
```
Comment (tyler-ball) [inline lib/rubocop/cop/chef/modernize/chef_14_resources.rb]: This list is different than what is in the example - is that intended?

File: config/cookstyle.yml (hunk 1 above)
Comment (tas50) [reply config/cookstyle.yml]: I need to further test them before I make them the default, but I want to remember that we should probably path limit these to speed up cookstyle

File: lib/rubocop/cop/chef/deprecation/easy_install.rb (hunk 2 above)
Comment (tas50) [reply lib/rubocop/cop/chef/deprecation/easy_install.rb]: Nope

File: lib/rubocop/cop/chef/deprecation/suggests_metadata.rb (hunk 3 above)
Comment (tas50) [reply lib/rubocop/cop/chef/deprecation/suggests_metadata.rb]: Fixed

File: lib/rubocop/cop/chef/modernize/chef_14_resources.rb (hunk 4 above)
Comment (tas50) [reply lib/rubocop/cop/chef/modernize/chef_14_resources.rb]: Fixed

**Pull request merged**